specification/dartLangSpec.tex

```diff
@@ -5028,6 +5078,1074 @@ \subsection{Mixin Application}
+\code{$E$<\List{T}{1}{k}>($e$)}
+where the type parameters of $E$ are \List{X}{1}{k}
+and the actual values of \List{T}{1}{k} are \List{t}{1}{k}
+(\ref{actualTypes}),
+and whose corresponding member name is $n$.
+Let $m$ be the member of $E$ that has the name $n$.
+Evaluation of $i$ proceeds by evaluating
+$e$ to an object $o$,
+evaluating and binding the actual arguments to the formal parameters
+(\ref{bindingActualsToFormals}),
+and finally executing $m$
+in a binding environment where \List{X}{1}{k} are bound to \List{t}{1}{k},
+\THIS{} is bound to $o$,
+and each formal parameter is bound to the corresponding actual argument.
+The value of $i$ is the value returned by the invocation of $m$.
+
+\LMHash{}%
+When $i$ is a conditional or composite member invocation,
+the static analysis and dynamic semantics is determined by
+member invocation desugaring
+(\ref{memberInvocations}).
+
+\rationale{%
+Note that a cascade (\ref{cascades})
+whose receiver is an extension application $a$ is a compile-time error.
+This is so because it implies that $a$ denotes an object, which is not true,
+and also because it would force each \synt{cascadeSection}
+to invoke a member of the same extension,
+which is unlikely to be desirable.%
+}
+
+
+\subsection{Implicit Invocation of an Instance Member of a Extension}
+\LMLabel{implicitExtensionInvocations}
+
+\LMHash{}%
+Instance members of an extension can be invoked implicitly
+(\commentary{without mentioning the name of the extension}),
+as if they were instance members of the receiver of
+the given member invocation.
+
+\commentary{%
+For instance, if \code{E<$T_1, T_2$>($e_1$).m<$T_3$>($e_2$)} is
+a correct explicit invocation of the instance member \code{m} of
+an extension \code{E},
+then \code{$e_1$.m<$T_3$>($e_2$)} may be
+a correct implicit invocation with the same meaning.
+In other words,
+the receiver $r$ of a member invocation can be implicitly replaced by
+an extension application with receiver $r$,
+if a number of requirements that are detailed below are satisfied.%
+}
+
+\rationale{%
+Implicit invocation is intended as the primary way to use extensions,
+with explicit invocation as a fallback
+in case the implicit invocation is an error, or
+the implicit invocation resolves to an instance member of
+a different extension than the intended one.%
+}
+
+\LMHash{}%
+An implicit extension member invocation occurs
+for a member invocation $i$
+(\ref{memberInvocations})
+with receiver $r$ and corresponding member name $m$ if{}f
+(1) the interface of the static type of $r$ does not have a member
+whose basename is the basename of $m$,
+and (2) there exists a unique most specific
+(\ref{extensionSpecificity})
+extension denoted by $E$ which is accessible
+(\ref{extensionAccessibility})
+and applicable
+(\ref{extensionApplicability})
+to $i$.
+
+\LMHash{}%
+In the case where no compile-time error occurs,
+$i$ is treated as $i'$, which is obtained from $i$ by
+replacing $r$ by \code{$E$($r$)}.
```

Comment:
by replacing $r$ by -> by replacing the leading $r$ by

(Just to cover the case where `r` occurs more than once in the invocation.)